src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol.java

```diff
@@ -49,16 +57,44 @@ protected JnlpProtocol(String secretKey, String slaveName) {
     public abstract String getName();
 
     /**
+     * Performs a handshake with the master and establishes a {@link Channel}.
+     *
+     * @param socket The established {@link Socket} connection to the master.
+     * @param channelBuilder The {@link ChannelBuilder} to use.
+     * @return The established channel if successful, else null.
+     * @throws IOException
+     */
+    @Nullable
+    public Channel establishChannel(Socket socket, ChannelBuilder channelBuilder) throws IOException {
+        DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream());
+        BufferedInputStream inputStream = new BufferedInputStream(socket.getInputStream());
+        if(performHandshake(outputStream, inputStream)) {
+            return buildChannel(socket, channelBuilder);
```

Comment:
And here we throw away the Buffer and all its read-ahead input causing the capabilities to be inferred as 0 and then the channel transport is configured without chunking and bamb

```
INFO: Protocol failed to establish channel
java.io.StreamCorruptedException: invalid stream header: 0A6CACED
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:806)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)
	at hudson.remoting.ChannelBuilder.makeTransport(ChannelBuilder.java:430)
	at hudson.remoting.ChannelBuilder.negotiate(ChannelBuilder.java:389)
	at hudson.remoting.ChannelBuilder.build(ChannelBuilder.java:310)
	at org.jenkinsci.remoting.engine.JnlpProtocol2.buildChannel(JnlpProtocol2.java:93)
	at org.jenkinsci.remoting.engine.JnlpProtocol.establishChannel(JnlpProtocol.java:79)
	at hudson.remoting.Engine.run(Engine.java:245)
```